### PR TITLE
First Release of DDQ Backend Mock

### PR DESCRIPTION
**Pull request by AbsentSemicolon**
Sets up having a mock for providing DDQ and example for further implementation and how a backend could work.

### CODE REVIEW

File: lib/ddq-backend-mock.js [hunk 1]
```diff
@@ -0,0 +1,304 @@
...
+                finalMessage.requeued = false;
+                this.writeRecord(finalMessage);
+
+                callback();
+            }
+        }
+
+
+        /**
+         * Starts polling for data. If data isfound it emits "data" so DDQ knows
```
Comment (lnagle) [inline lib/ddq-backend-mock.js]: Need a space in 'isfound'.

File: package.json [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+  "private": false,
+  "repository": {
+      "type": "git",
+      "url": "git+https://github.com/tests-always-included/ddq-backend-mock.git"
+  },
+  "scripts": {
+    "test": "eslint ."
+  },
+  "version": "1.0.0"
+}
```
Comment (lnagle) [inline package.json:31]: Indentation is off throughout.

File: .eslintrc.yaml [hunk 3]
```diff
@@ -0,0 +1,232 @@
+parserOptions:
+    ecmaVersion: 6
+env:
+    es6: true
+    jasmine: true
+    node: true
+extends:
+    - eslint:recommended
```
Comment (fidian) [inline .eslintrc.yaml:8]: Change to

   -
        eslint: recommended

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -0,0 +1,37 @@
+Contributing
+============
+
+A guide to contributing for DDQ Backend Mock.
+
+## Getting started
+
+1. Fork the repo. There is a very simple way of doing this on GitHub.
```
Comment (fidian) [inline CONTRIBUTING.md]: Link to alternate documentation, especially GitHub's documentation.  Do not write it up here.

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+    * Following steps may not be needed if forked properly, but in case you can't get to the upstream server.
+        1. Configure the upstream server by running `git remote add upstream <repo address>` or editing your `.git/config` under `[remote "upstream"]`, changing `url`.
+3. Create a branch off master `git checkout -b <branch name>`. You don't want to do any work off master as this is where you will pull in upstream changes and merge them into your branches. Also is a good idea to push this remotely `git push -u origin <branch name>`.
```
Comment (fidian) [inline CONTRIBUTING.md]: "2", not "3"

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+4. Write your code as normal.  Be sure to follow the [Tests-Always-Included Style Guide](https://tests-always-included.github.io/style-guide/)
```
Comment (fidian) [inline CONTRIBUTING.md]: "3", not "4".

File: CONTRIBUTING.md [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+
+## Pull Requests
+
+Before creating a pull request it's a good idea to pull in the upstream changes, just in case someone has created a Pull Request before you and it has been merged in.
```
Comment (fidian) [inline CONTRIBUTING.md]: Pull Request is not a Proper Noun and We don't really prefer to speak like Winnie the Pooh.

File: CONTRIBUTING.md [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+
+## Getting started
+
+1. Fork the repo. There is a very simple way of doing this on GitHub.
+    * Following steps may not be needed if forked properly, but in case you can't get to the upstream server.
+        1. Configure the upstream server by running `git remote add upstream <repo address>` or editing your `.git/config` under `[remote "upstream"]`, changing `url`.
+3. Create a branch off master `git checkout -b <branch name>`. You don't want to do any work off master as this is where you will pull in upstream changes and merge them into your branches. Also is a good idea to push this remotely `git push -u origin <branch name>`.
+4. Write your code as normal.  Be sure to follow the [Tests-Always-Included Style Guide](https://tests-always-included.github.io/style-guide/)
+
+## Pull Requests
```
Comment (fidian) [inline CONTRIBUTING.md]: This does not need a separate section.  Omit this line.

File: CONTRIBUTING.md [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+
+Before creating a pull request it's a good idea to pull in the upstream changes, just in case someone has created a Pull Request before you and it has been merged in.
+
+1. Checkout the master branch `git checkout master`.
```
Comment (fidian) [inline CONTRIBUTING.md]: Can we just link somewhere instead of writing these instructions up all the time?

File: CONTRIBUTING.md [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+2. Run `git fetch upstream`.
+    *  This will pull in all the changes from the remote ddq-backend-mock repo master branch.
+3. Run `git merge upstream/master`.
+    *  This will merge the changes into master branch.
+4. Checkout your feature branch.
+5. Merge master into your branch. `git merge master <branch name>`
+    * If everything merges in fine then you're good to create a Pull Request, otherwise fix merge conflicts or other issues.
+6. We suggest squashing commits, if able, and creating a Pull Request on GitHub, but this has been problematic when merge conflicts have been present. The other way is to create a patch. Follow the instructions below to create a patch.
```
Comment (fidian) [inline CONTRIBUTING.md]: Not necessary, I'll happily squash them for you.  Plus we can link somewhere.

File: lib/ddq-backend-mock.js [hunk 11]
```diff
@@ -0,0 +1,304 @@
+"use strict";
+
+/**
+ * @typeDef {Object} DdqBackend~record
```
Comment (fidian) [inline lib/ddq-backend-mock.js]: Lowercase d

File: lib/ddq-backend-mock.js [hunk 12]
```diff
@@ -0,0 +1,304 @@
...
+ */
+
+
+/**
+ * What the backend sets up and sends to the frontend.
+ *
+ * @typeDef {Object} Ddq~wrappedMessage
+ * @property {Ddq~config} config
+ * @property {Function} heartbeat
+ * @property {Function} heartbeatKill
```
Comment (fidian) [inline lib/ddq-backend-mock.js]: Delete me.

File: lib/ddq-backend-mock.js [hunk 13]
```diff
@@ -0,0 +1,304 @@
...
+
+
+        /**
+         * Want to make an instance id and return if we need to.
+         *
+         * @param {Function} callback
+         */
+        getId(callback) {
+            if (this.instanceId) {
+                callback(this.instanceId);
```
Comment (fidian) [inline lib/ddq-backend-mock.js]: The callback should be a standard Node-style callback, with an optional error as the first parameter.

File: lib/ddq-backend-mock.js [hunk 14]
```diff
@@ -0,0 +1,304 @@
...
+            callback();
+        });
+    }
+
+
+    /**
+     * @param {Object} ddqBackendInstance
+     * @param {string} recordId
+     */
+    function heartbeatKill(ddqBackendInstance, recordId) {
```
Comment (fidian) [inline lib/ddq-backend-mock.js]: This should not exist.

File: lib/ddq-backend-mock.js [hunk 15]
```diff
@@ -0,0 +1,304 @@
...
+         */
+        getId(callback) {
+            if (this.instanceId) {
+                callback(this.instanceId);
+            } else {
+                crypto.randomBytes(16, (err, id) => {
+                    if (err) {
+                        callback();
+                    }
+                    this.instanceId = id.toString("hex");
```
Comment (fidian) [inline lib/ddq-backend-mock.js]: If we are using an in-memory data store, why do we care about the instance ID?  This additional complexity should be removed.

File: lib/ddq-backend-mock.js [hunk 16]
```diff
@@ -0,0 +1,304 @@
...
+         * For times when polling needs to pause so no more messages are sent.
+         */
+        pausePolling() {
+            timers.clearInterval(this.poller);
+            this.poller = null;
+        }
+
+
+        /**
+         * Gets the stored data or a particular one beind searched for.
```
Comment (fidian) [inline lib/ddq-backend-mock.js]: The second half of the sentence is not reflected in the code below.

File: lib/ddq-backend-mock.js [hunk 17]
```diff
@@ -0,0 +1,304 @@
...
+    function heartbeat(ddqBackendInstance, recordId, callback) {
+        // Need to get the ID asynchronously.  :-(
+        ddqBackendInstance.getId((instanceId) => {
+            var record;
+
+            if (!instanceId) {
+                ddqBackendInstance.emit(new Error("Could not make instance id"));
+            }
+
+            record = ddqBackendInstance.getRecord(recordId);
```
Comment (fidian) [inline lib/ddq-backend-mock.js]: What if this fails?

File: lib/ddq-backend-mock.js [hunk 18]
```diff
@@ -0,0 +1,304 @@
+"use strict";
+
+/**
+ * @typeDef {Object} DdqBackend~record
+ * @property {(string|null)} heartbeatTime
```
Comment (fidian) [inline lib/ddq-backend-mock.js]: This should just be a Date object.

File: lib/ddq-backend-mock.js [hunk 19]
```diff
@@ -0,0 +1,304 @@
...
+                // eslint-disable-next-line no-bitwise
+                file = files[keys[keys.length * Math.random() << 0]];
+
+                if (file && !file.isProcessing) {
+                    files[file.id].isProcessing = true;
+                    this.writeRecord(file);
+
+                    return {
+                        heartbeat: heartbeat.bind(null, this, file.id),
+                        heartbeatKill: heartbeatKill.bind(null, this.file.id),
```
Comment (fidian) [inline lib/ddq-backend-mock.js]: Delete

File: lib/ddq-backend-mock.js [hunk 20]
```diff
@@ -0,0 +1,304 @@
...
+         */
+        getWrappedMessage() {
+            var file, files, keys;
+
+            files = this.readData();
+            keys = Object.keys(files);
+
+            if (keys.length) {
+                // eslint-disable-next-line no-bitwise
+                file = files[keys[keys.length * Math.random() << 0]];
```
Comment (fidian) [inline lib/ddq-backend-mock.js]: Why bitwise shift?  Just to cast to an integer?

```
Math.floor(keys.length * Math.random())
```

However, I think you should just use a shuffle.  http://stackoverflow.com/questions/962802/is-it-correct-to-use-javascript-array-sort-method-for-shuffling  Then iterate over the list and emit ALL of the datas or until you're paused.
Maybe the poll amount (max number tried) could be another config option that could match the max number being processed?

File: lib/ddq-backend-mock.js [hunk 21]
```diff
@@ -0,0 +1,304 @@
...
+
+
+        /**
+         * Looks to see if there are any messages in the storeData and creates
+         * the wrapped message.
+         *
+         * @return {(Ddq~wrappedMessage|null)}
+         */
+        getWrappedMessage() {
+            var file, files, keys;
```
Comment (fidian) [inline lib/ddq-backend-mock.js]: Not really files.

File: lib/ddq-backend-mock.js [hunk 22]
```diff
@@ -0,0 +1,304 @@
...
+                if (file && !file.isProcessing) {
+                    files[file.id].isProcessing = true;
+                    this.writeRecord(file);
+
+                    return {
+                        heartbeat: heartbeat.bind(null, this, file.id),
+                        heartbeatKill: heartbeatKill.bind(null, this.file.id),
+                        message: file.message,
+                        remove: remove.bind(null, this, file.id),
+                        requeue: requeue.bind(null, this, file.id)
```
Comment (fidian) [inline lib/ddq-backend-mock.js]: On remove or requeue, make sure to do another immediate poll.  Only one needs to be sent at that time, not all of them.

File: lib/ddq-backend-mock.js [hunk 23]
```diff
@@ -0,0 +1,304 @@
...
+         * Looks to see if there are any messages in the storeData and creates
+         * the wrapped message.
+         *
+         * @return {(Ddq~wrappedMessage|null)}
+         */
+        getWrappedMessage() {
+            var file, files, keys;
+
+            files = this.readData();
+            keys = Object.keys(files);
```
Comment (fidian) [inline lib/ddq-backend-mock.js]: I would suggest filtering the keys to eliminate entries that are being processed.

We think that the data store should be a Map, not an Object.

File: lib/ddq-backend-mock.js [hunk 24]
```diff
@@ -0,0 +1,304 @@
...
+
+            if (keys.length) {
+                // eslint-disable-next-line no-bitwise
+                file = files[keys[keys.length * Math.random() << 0]];
+
+                if (file && !file.isProcessing) {
+                    files[file.id].isProcessing = true;
+                    this.writeRecord(file);
```
Comment (fidian) [inline lib/ddq-backend-mock.js]: You're dealing with objects.  Be careful.  The line above already did what you wanted.

File: lib/ddq-backend-mock.js [hunk 25]
```diff
@@ -0,0 +1,304 @@
...
+            }
+        }
+
+
+        /**
+         * Starts polling for data. If data isfound it emits "data" so DDQ knows
+         * to take it's action for finding a piece of data.
+         */
+        startPolling() {
+            this.poller = timers.setInterval(() => {
```
Comment (fidian) [inline lib/ddq-backend-mock.js]: No, not intervals!  Aaahahhhahaa!

File: lib/ddq-backend-mock.js [hunk 26]
```diff
@@ -0,0 +1,304 @@
...
+                callback(new Error("Could not send"));
+            } else {
+                finalMessage = {};
+                finalMessage.heartbeatTime = null;
+                finalMessage.id = crypto.createHash("sha256").update(message).digest("hex");
+                finalMessage.isProcessing = false;
+                finalMessage.message = message;
+                finalMessage.owner = null;
+                finalMessage.requeued = false;
+                this.writeRecord(finalMessage);
```
Comment (fidian) [inline lib/ddq-backend-mock.js]: Note, this STAYS but you could just replace this line with `files[finalMessage.id] = finalMessage`;

File: package.json [hunk 27]
```diff
@@ -0,0 +1,29 @@
...
+    "eslint": "^2.10.2"
+  },
+  "engines": {
+    "node": "*"
+  },
+  "homepage": "https://github.com/tests-always-included/ddq-backend-mock#readme",
+  "license": "MIT",
+  "main": "lib/index.js",
+  "name": "ddq-backend-mock",
+  "private": false,
```
Comment (fidian) [inline package.json]: Drop this line

File: package.json [hunk 28]
```diff
@@ -0,0 +1,29 @@
+{
+  "author": "Christopher Tetreault",
+  "bugs": {
+    "url": "https://github.com/tests-always-included/ddq-backend-mock/issues"
+  },
+  "dependencies": {
+    "ddq": "git://github.com/AbsentSemicolon/ddq.git#release-1"
```
Comment (fidian) [inline package.json]: Drop

File: package.json [hunk 29]
```diff
@@ -0,0 +1,29 @@
...
+  },
+  "dependencies": {
+    "ddq": "git://github.com/AbsentSemicolon/ddq.git#release-1" 
+  },
+  "description": "DeDuplicated Queue Backend Mock",
+  "devDependencies": {
+    "eslint": "^2.10.2"
+  },
+  "engines": {
+    "node": "*"
```
Comment (fidian) [inline package.json]: Isn't this >=4

**Review by fidian — CHANGES_REQUESTED** [review]

File: lib/ddq-backend-mock.js (hunk 1 above)
Comment (AbsentSemicolon) [reply lib/ddq-backend-mock.js]: Spaced

File: package.json (hunk 2 above)
Comment (AbsentSemicolon) [reply package.json:31]: Shouldn't let package.json do most of the indenting.

File: .eslintrc.yaml (hunk 3 above)
Comment (AbsentSemicolon) [reply .eslintrc.yaml:8]: This is supposed to be a string so it needs to stay like that.

File: CONTRIBUTING.md (hunk 4 above)
Comment (AbsentSemicolon) [reply CONTRIBUTING.md]: Added a link to Githubs doc on forking.

File: CONTRIBUTING.md (hunk 5 above)
Comment (AbsentSemicolon) [reply CONTRIBUTING.md]: Think I took something out of here and didn't update the numbers.

File: CONTRIBUTING.md (hunk 6 above)
Comment (AbsentSemicolon) [reply CONTRIBUTING.md]: Updated

File: lib/ddq-backend-mock.js (hunk 11 above)
Comment (AbsentSemicolon) [reply lib/ddq-backend-mock.js]: ok.

File: lib/ddq-backend-mock.js (hunk 18 above)
Comment (AbsentSemicolon) [reply lib/ddq-backend-mock.js]: ok.

File: lib/ddq-backend-mock.js (hunk 12 above)
Comment (AbsentSemicolon) [reply lib/ddq-backend-mock.js]: Yep, gone.

File: lib/ddq-backend-mock.js (hunk 14 above)
Comment (AbsentSemicolon) [reply lib/ddq-backend-mock.js]: Removed

File: lib/ddq-backend-mock.js (hunk 24 above)
Comment (AbsentSemicolon) [reply lib/ddq-backend-mock.js]: Right, this shouldn't be there.

File: lib/ddq-backend-mock.js (hunk 19 above)
Comment (AbsentSemicolon) [reply lib/ddq-backend-mock.js]: Deleted

File: lib/ddq-backend-mock.js (hunk 25 above)
Comment (AbsentSemicolon) [reply lib/ddq-backend-mock.js]: AHHHHH!!!! This has been replaced with setTimeout.

File: package.json (hunk 28 above)
Comment (AbsentSemicolon) [reply package.json]: Dropped

File: package.json (hunk 29 above)
Comment (AbsentSemicolon) [reply package.json]: Made it so we support node great or equal to 4.

File: package.json (hunk 27 above)
Comment (AbsentSemicolon) [reply package.json]: Dropped

File: lib/ddq-backend-mock.js (hunk 26 above)
Comment (AbsentSemicolon) [reply lib/ddq-backend-mock.js]: Yeah no need for write record if we are dealing with object we can just edit. We aren't protecting them or anything.

File: lib/ddq-backend-mock.js (hunk 16 above)
Comment (AbsentSemicolon) [reply lib/ddq-backend-mock.js]: Was the original thought behind this function.

File: lib/ddq-backend-mock.js (hunk 22 above)
Comment (AbsentSemicolon) [reply lib/ddq-backend-mock.js]: Making a call to get another message as soon as we requeue and remove.

File: lib/ddq-backend-mock.js (hunk 21 above)
Comment (AbsentSemicolon) [reply lib/ddq-backend-mock.js]: This uses records now.

File: lib/ddq-backend-mock.js (hunk 15 above)
Comment (AbsentSemicolon) [reply lib/ddq-backend-mock.js]: Took out owner, not really needed.

File: lib/ddq-backend-mock.js (hunk 13 above)
Comment (AbsentSemicolon) [reply lib/ddq-backend-mock.js]: This was taken out.

File: lib/ddq-backend-mock.js (hunk 17 above)
Comment (AbsentSemicolon) [reply lib/ddq-backend-mock.js]: Sending error back as the front end is expecting.

File: CONTRIBUTING.md (hunk 7 above)
Comment (AbsentSemicolon) [reply CONTRIBUTING.md]: Fixed.

File: CONTRIBUTING.md (hunk 8 above)
Comment (AbsentSemicolon) [reply CONTRIBUTING.md]: Took this out.

File: lib/ddq-backend-mock.js (hunk 20 above)
Comment (AbsentSemicolon) [reply lib/ddq-backend-mock.js]: Found a way to randomize the array and just pull the first one for processing. Also using map to populate an array with ones which are not processing.

File: lib/ddq-backend-mock.js (hunk 23 above)
Comment (AbsentSemicolon) [reply lib/ddq-backend-mock.js]: Made this a map.

File: CONTRIBUTING.md (hunk 9 above)
Comment (AbsentSemicolon) [reply CONTRIBUTING.md]: Linked to an article about pulling in upstream changes.

File: CONTRIBUTING.md (hunk 10 above)
Comment (AbsentSemicolon) [reply CONTRIBUTING.md]: linked to an article about pull requests in Github.

**Review by fidian — APPROVED** [review]

**Pull request merged**